### PR TITLE
Block Bindings: Prioritize existing placeholder over bindingsPlaceholder

### PR DESCRIPTION
**Pull request by zaguiini**
Supersedes #65154.

## What?

Respect existing `placeholder` attribute when it exists over the one provided by the Block Bindings API.

## Why?

We're setting a more helpful placeholder if the block has bindings. While that's awesome and works as intended most of the time, if the block already has an existing placeholder, that value is lost, but it should've been prioritized.

## How?

Prioritize `placeholder` in the applicable places (`placeholder` and `aria-label`.) We're using an existing prop called `data-custom-placeholder` to understand if the placeholder has been customized.

## Testing Instructions

Register a new post meta field on the server:

```php
add_action(
	'init',
	function () {
		register_meta(
			'post',
			'details',
			array(
				'type'         => 'string',
				'show_in_rest' => true,
				'single'       => true,
				'default'      => '',
			)
		);
	}
);
```

In Gutenberg...
1. define a placeholder for a Paragraph block, then bind the attribute.
2. create a new Paragraph block and bind the attribute.

The result should match the screenshot below.

Side note: The 2nd paragraph is empty because the value returned from the server is an empty string. I'm not sure that's intended, but that's what we have right now in `trunk`. 🤔 I'm also unable to edit the value, which is weird considering recently we've fixed a similar bug.

## Screenshots or screencast 

image

### CODE REVIEW

File: packages/block-editor/src/components/rich-text/index.js [hunk 1]
```diff
@@ -406,7 +408,11 @@ export function RichTextWrapper(
 				aria-readonly={ shouldDisableEditing }
 				{ ...props }
 				aria-label={
-					bindingsPlaceholder || props[ 'aria-label' ] || placeholder
```
Comment (zaguiini) [inline packages/block-editor/src/components/rich-text/index.js:424]: @gziolo, this change broke the `Cover` block tests.

But this situation is confusing for me: what is the priority here?
- Do we always want to respect the `aria-label`?
- Do we want to have the custom placeholder as the `aria-label`?
- Why would we want the bindings placeholder to take priority over the `aria-label`?

I'd love it if you could clarify this to me.

Comment (gziolo) [reply packages/block-editor/src/components/rich-text/index.js:424]: That's probably this code:

Screenshot 2024-09-12 at 13 44 43

If we have a bound block, we rather need to show something else. However, it looks like `aria-label` should be the original one when the attribute isn't connected.

Comment (zaguiini) [reply packages/block-editor/src/components/rich-text/index.js:424]: Gotcha @gziolo. It seems weird to me that we want to display the bindings placeholder as `aria-label` considering a custom placeholder exists, but I reverted the `aria-label` changes and this PR should be good to go. Please take another look!

**Comment by SantosGuillamot** [comment]
At first glance, I would try to avoid adding a new prop to the rich text component. I still have to take a deeper look at the code, but maybe we can read the `placeholder` attribute or move the `bindingsPlaceholder` logic to another place.

**Comment by zaguiini** [comment]
>At first glance, I would try to avoid adding a new prop to the rich text component.

It's an already existing prop 😅

>maybe we can read the placeholder attribute

We'll need to extract the default somewhere else, but even then, the default varies by element type. It's a bit tricky, I've thought about that for a few minutes.

**Comment by SantosGuillamot** [comment]
> It's an already existing prop

Sorry, I didn't realize that. Anyway, it doesn't feel intuitive to me, but I'd love to know other opinions.

> We'll need to extract the default somewhere else, but even then, the default varies by element type. It's a bit tricky, I've thought about that for a few minutes.

I was thinking about changing [this conditional](https://github.com/WordPress/gutenberg/blob/782dd99252e8bbbe335d453c82effacfc465f819/packages/block-editor/src/components/rich-text/index.js#L199-L201) to something like this:

```js
bindingsPlaceholder:
	adjustedValue?.length > 0 || attributes?.placeholder
		? undefined
		: _bindingsPlaceholder,
```

I am not sure if the block attributes are already available in the rich text component, but we have the `clientId` so we could always call `getBlockAttributes`.

I quickly tested and it seems to work, although it would need a double-check.

---
By the way, I believe we should combine the `useSelect` with [this other one](https://github.com/WordPress/gutenberg/blob/782dd99252e8bbbe335d453c82effacfc465f819/packages/block-editor/src/components/rich-text/index.js#L128-L165). But I can work on that in another pull request.

**Comment by zaguiini** [comment]
>I was thinking about changing [this conditional](https://github.com/WordPress/gutenberg/blob/782dd99252e8bbbe335d453c82effacfc465f819/packages/block-editor/src/components/rich-text/index.js#L199-L201) to something like this:

Would that work, considering the placeholder attribute is always set? That's why the `data-custom-placeholder` prop exists: to differentiate the default placeholder from the block and the placeholder attribute.

**Comment by SantosGuillamot** [comment]
> Would that work, considering the placeholder attribute is always set?

If I am not mistaken, the `placeholder` **block attribute** is not always set. It is different from the `placeholder` **RichText prop**. You can have a paragraph without the `placeholder` attribute defined but it passes the "Type to..." text as placeholder to the `RichText` component.

**Comment by SantosGuillamot** [comment]
Maybe the logic could be moved [here](https://github.com/WordPress/gutenberg/pull/65220/files#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44R337-R339), but I believe this order still applies: Placeholder attribute -> Bindings placeholder -> Placeholder prop.

**Comment by SantosGuillamot** [comment]
In the latest two commits, I've pushed a different alternative to solve the issue. Basically, it checks the `placeholder` block attribute before overriding it with the bindings placeholder.

Additionally, I slightly modified the logic for `aria-label` to not rely on the bindings placeholder. I believe it is clearer this way and it provides more flexibility to decide the message there. Aria-label and placeholder are also different for normal empty paragraphs.

Let me know what you think 🙂

**Comment by zaguiini** [comment]
Looks great to me @SantosGuillamot 🚀

**Comment by SantosGuillamot** [comment]
I have rebased and solved the conflicts in this branch.

File: packages/block-editor/src/components/rich-text/index.js [hunk 2]
```diff
@@ -191,6 +187,19 @@ export function RichTextWrapper(
 					args: relatedBinding.args,
 				} );
 
+			// Don't modify placeholders if value is not empty.
+			if ( adjustedValue.length > 0 ) {
+				return {
+					disableBoundBlock: _disableBoundBlock,
```
Comment (gziolo) [inline packages/block-editor/src/components/rich-text/index.js:193]: Interesting, should it explicitly return `null` for `bindingsPlaceholder` and `bindingsLabel` with some more detailed explanation? 

Alternatively, would the following work and made it easier to follow:

```suggestion
					disableBoundBlock: _disableBoundBlock,
					bindingsLabel: props[ 'aria-label' ] || placeholder,
```

Comment (gziolo) [reply packages/block-editor/src/components/rich-text/index.js:193]: A similar trick might also work for `bindingsPlaceholder`.

**Review by gziolo — APPROVED** [review]
I left only feedback related to code readability. This fix is looking great. I think we should include in the plugin release and backport also to WP 6.7 release.

File: packages/block-editor/src/components/rich-text/index.js (hunk 2 above)
Comment (SantosGuillamot) [reply packages/block-editor/src/components/rich-text/index.js:193]: I've explicitly return `null` as suggested. I kind of like having the conditionals outside of the bindings logic. It makes it clear when the `placeholder` and the `aria-label` are defined:

* https://github.com/WordPress/gutenberg/blob/0027219426d4db3b7f5c38ee7eef844724024a35/packages/block-editor/src/components/rich-text/index.js#L372
* https://github.com/WordPress/gutenberg/blob/0027219426d4db3b7f5c38ee7eef844724024a35/packages/block-editor/src/components/rich-text/index.js#L442-L444

Having said that, I don't have a strong opinion so I'm happy to change that.

**Comment by kevin940726** [comment]
I just cherry-picked this PR to the release/19.3 branch to get it included in the next release: 3640694f515

**Pull request merged**